src/modules/linux-micro/kmod/kmod.c

```diff
@@ -0,0 +1,475 @@
+    }
+
+    if (!modlist) {
+        SOL_INF("No modules found for alias: '%s'", alias);
+        return -ENOENT;
+    }
+
+    kmod_list_foreach(itr, modlist) {
+        struct kmod_module *mod;
+        int state, err;
+
+        mod = kmod_module_get_module(itr);
+        state = kmod_module_get_initstate(mod);
+
+        switch (state) {
+        case KMOD_MODULE_BUILTIN:
+            SOL_DBG("Module '%s' is builtin", kmod_module_get_name(mod));
+            break;
+
+        case KMOD_MODULE_LIVE:
+            SOL_DBG("Module '%s' is already loaded", kmod_module_get_name(mod));
+            break;
+
+        default:
+            err = kmod_module_probe_insert_module(mod, probe_flags,
+                NULL, NULL, NULL, NULL);
+
+            if (err == 0)
+                SOL_INF("Inserted module '%s'", kmod_module_get_name(mod));
+            else if (err == KMOD_PROBE_APPLY_BLACKLIST)
+                SOL_INF("Module '%s' is blacklisted", kmod_module_get_name(mod));
+            else {
+                SOL_WRN("Failed to insert '%s': %s", kmod_module_get_name(mod),
+                    sol_util_strerrora(errno));
+                r = err;
+            }
+        }
+
+        kmod_module_unref(mod);
+    }
+
+    kmod_module_unref_list(modlist);
+    return r;
+}
+
+static int
+kmod_apply_file(struct kmod_ctx *kmod, int fd)
+{
+    int err = 0, ret;
+    struct sol_file_reader *reader;
+    struct sol_str_slice content = SOL_STR_SLICE_EMPTY;
+    struct sol_str_slice *itr;
+    uint16_t idx;
+    struct sol_vector lines;
+
+    reader = sol_file_reader_fd(fd);
+    if (!reader) {
+        SOL_ERR("Could not open config file");
+        return err;
+    }
+
+    content = sol_file_reader_get_all(reader);
+    if (content.len == 0)
+        goto finish;
+
+    lines = sol_util_str_split(content, "\n", 0);
+    SOL_VECTOR_FOREACH_IDX (&lines, itr, idx) {
+        struct sol_str_slice slice;
+
+        slice = sol_str_slice_remove_leading_whitespace(*itr);
+        if (*slice.data == '#' || *slice.data == ';' || *slice.data == '\0')
+            continue;
+
+        slice = sol_str_slice_remove_trailing_whitespace(slice);
+        ret = kmod_apply_value(kmod, slice);
+        if (ret < 0)
+            err = ret;
+    }
+    sol_vector_clear(&lines);
+
+finish:
+    sol_file_reader_close(reader);
+    return err;
+}
+
+static int
+kmod_apply_filename(struct kmod_ctx *kmod, int dfd, const char *file_name)
+{
+    int fd;
+
+    if (*file_name == '/' || dfd < 0)
+        fd = open(file_name, O_RDONLY | O_CLOEXEC);
+    else
+        fd = openat(dfd, file_name, O_RDONLY | O_CLOEXEC);
+
+    if (fd < 0) {
+        SOL_DBG("Could not open: %s", file_name);
+        return 0;
+    }
+
+    return kmod_apply_file(kmod, fd);
+}
+
+static int
+kmod_conf_cmp(const void *data1, const void *data2)
+{
+    return strcmp((const char *)data1, (const char *)data2);
+}
+
+static void
+read_directory_to_vector(DIR *dir, struct sol_ptr_vector *vec)
```

Comment:
Not for this patch, but maybe create an utility function to do this, filtering by file extension? Or, better yet, a function that walks a directory path and calls a callback function for each name a `fnmatch()` pattern match? This would simplify the `kmod`, `sysctl`, and possibly other modules as well.